design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,503 @@
+pods will calculate an allocatable ip from the `NodeIPPool` status and
+retry the process on conflict
+
+Handlers will check NNCEs at Reconcile and if the ip has being already assigned
+it will re-use that and update `NodeIPPool` status with it, this way the
+`NodeIPPool` can be reset but IPs are kept.
+
+To optimize on storage and lookup instead of using a list of allocated offset
+it's possible to use a bitmap, the OVN project use that approach to store the
+offsets on a range [1], a snapshot of the bitmap can be stored as base64 at the
+`NodeIPPool` status.
+
+For the scale down scenario, the current kubernetes-nmstate architecture
+contains a controller for node that will Reconcile on node deletion, but since
+the nodes may be drained before deletion the pod running the controller
+will no longer be running, this means that the controller may need to run at
+control-plane nodes and watch for all the nodes.
+
+[1] https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ipallocator/allocator.go
+
+#### Baremetal IPI
+
+The `image-customize-controller` [1] is a golang controller so it would be quite
+easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
+`networkPolicy`.
+
+[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79
+
+### Risks and Mitigations
+
+If a bug prevents IPs deallocation the ippool can be exhausted and end up
+with network reconfiguration or scaling up failures. To fix those situations
+the field `status.allocated-offsets` at `NodeIPPool` can be set with an
+empty list triggering a pool reconstruction, bringing back the
+stale IPs. This risk gets higher for shorter IPPools.
+
+If the `NodeIPPool` is exhausted the values from the `spec` fields can be
+changed to add more free IPs to the pool. Doing so involves that handlers will
+reconstruct `status.allocated-offsets` with the already reserved IPs at NNCE.
+
+### Drawbacks
+
+Updating the pool is distributed across nodes by the nmstate-handler pod, this
+means that N nodes will try to update `NodeIPPool` and retry on conflict. Is
+not expected the IPAM mechanism will run very often so this should not be a
+problem.
+
+Debugging the allocated offset can be a challange since it involve a convertion
+from the offset to the IP is pointing too depending on `NodeIPPool` spec.
+
+Fixing the "exhausted" issue updating the NodeIPPool involves a lot of logic to
+reconstruct the `allocated-offsets` that has to be carefully implemented and
+tested, what alternate would be to just delete `NodeIPPool` and create new one,
+so `NodeIPPool` is a read-only resource.
+
+The baremetal IPI Nmstate does an offline NetworkManager keyfile generation so
+there is no network `currentState` that nmpolicy can reference to, so using
+nmpolicy just to reference the ippools may be too overkill.
+
+If in the future nmstatectl is included at FCOS then the `networkPolicy` will
```

Comment:
What's an FCOS ? (link to the fedora coreOS page please).

Reply:
Removing baremetal IPI reference.